### ISSUE TITLE
Perform batch RPC requests to ethereum node

### ISSUE THREAD

**Issue by Uxio0** [ethereum/web3.py#832]
* Version: 4.2.0
* Python: 3.6
* OS: osx/linux/win

### What was wrong?

There's currently not support for batch requests, when using `HTTPProvider`. Even using IPC/WS connectors I think a speed up could be possible using batch requests.

### How can it be fixed?

Implementing batch requests

**Comment by carver** [comment]
I don't understand what you mean by batching, here. IPC is a streaming connection. Can you elaborate?

**Comment by pipermerriam** [comment]
@carver web3js at least historically had a feature for asynchronous request batching via an API that let you stack up some JSON-RPC requests, and then submit them as a batch.  The requests would be issued concurrently.  (feature may still exist, not sure).

I'm good with exploring this alongside #657

**Comment by carver** [comment]
> web3js at least historically had a feature for asynchronous request batching via an API that let you stack up some JSON-RPC requests, and then submit them as a batch

Right, that makes sense in web3js, because the HTTP request to your node has a lot of overhead (HTTP headers) for each message.  But there is roughly 0 overhead for each message in an IPC stream. So I'm skeptical that it will provide any benefit in that context.

I'm less sure about the internals of websockets. There might be some benefit there.

Batching over an HTTP connection is likely to have benefits. :+1:

**Comment by Uxio0** [comment]
Hi @carver,

I agree about that, but:

- I'm talking about HTTP, not in every environment is possible to use IPC/WS.
- Even using IPC/WS, it's possible that Geth/Parity and other nodes have optimizations depending on if they are receiving a batch request or multiple streamed requests. At least for me is faster to send batchs than to send synchronous IPC requests (python 3.6 at least with web3 4.2.0).

Hope I have explained myself better :)

**Comment by pipermerriam** [comment]
@carver having now thought about this for a few minutes, I think this is going to be *easy* if we build it on top of #657 since everything under `web3.async.*` will be coroutine based and thus, batching them to run concurrently (in the asyncio version of concurrency) should be really easy to expose .

Here's some imagined API

```python
w3 = Web3(...)

batch = web3.async.create_batch()
batch.eth.getBalance(addr_a)
batch.eth.getBalance(addr_b)
batch.eth.getBalance(addr_c)
bal_a, bal_b, bal_c = await batch.execute()
```

**Comment by dangell7** [comment]
Interesting idea. What happens when one tries to send multiple tranascations? Cuz in my system, I have to wait till the first transaction clears before send the next. Prob because getTransactionCount doesn’t take the pending transaction into consideration. I didn’t really dive into the problem, but wanted to bring it up.

**Comment by pipermerriam** [comment]
@dangell7 good point.  We *may* choose to start with a minimal initial implementation which only allows batching requests that are *read-only* or that don't modify state.

**Comment by boneyard93501** [comment]
@pipermerriam anybody working on this? it's not all that far off the poc i did a while ago. might be able to resurrect that without too many changes as a wip starting point.

**Comment by carver** [comment]
@boneyard93501 not that I've seen

**Comment by boneyard93501** [comment]
@carver cool. if you don't mind, i'll take a stab at it starting wednesday.

**Comment by pipermerriam** [comment]
@boneyard93501 are you looking at starting into the `web3.async` API?  If not, can you provide a proposal for what you're looking to build?  Basic API and a basic description of how it'll work.

**Comment by boneyard93501** [comment]
@pipermerriam @carver 
if a solution is supposed to be provider agnostic, i think a thread-based pub-sub solution is probably the most straight forward (and could be wrapped in an event-loop, if needed). quick and dirty outline, untested, unoptimized/un-toolzed:
``` python
import time
import datetime
import queue
import threading


WORKER_TIMEOUT = 5.0


def batch_request_handler(rpc_request_queue, provider, response_processing_fn):
    '''  '''
    last_job = None
    while 1:
        try:
            if (datetime.datetime.utcnow() - last_job).total_seconds() > WORKER_TIMEOUT:
                response_processing_fn(('','', 'TimeoutError'))
                break
            job = rpc_request_queue.get()
            if job is None:
                break
            last_job = datetime.datetime.utcnow()
            try:
                result = provider._make_request(job)
                response_processing_fn((job,result, ''))
            except Exception as exc:
                response_processing_fn((job,'',exc))
        except (KeyboardInterrupt, SystemExit):
            pass
        finally:
            rpc_request_queue.task_done()
        time.sleep(0.1)


def batch_request(provider, payload, response_processing_fn, max_request_threads=10, max_request_q=200):
    '''
        could change provider to provider_type and then create provider in the worker to
        avoid any possible thread-safety issues now of down the road.
        also should take threading.Event, adjust worker while condition accordingly, so user can kill it all at will
   '''

    rpc_reqeuest_queue = queue.Queue()

    request_args = (rpc_reqeuest_queue, provider, response_processing_fn)
    req_threads = []

    for i in range(max_request_threads):
        t = threading.Thread(target=batch_request_handler, args=request_args)
        t.setDaemon(True)  # we don't really need it since we're joining but it helps to avoid zombies
        t.name = f'batch request thread {i}'
        req_threads.append(t)

    [t.start() for t in req_threads]

    for job in payload:
        rpc_reqeuest_queue.put(job)

    time.sleep(WORKER_TIMEOUT + 1.0)
    [t.join() for t in req_threads]

 

# user side
class SuperSimplClientBatchProcessingClass:
    def __init__(self, batch_size):
        ''' '''
        self.expected_results = batch_size
        self.jobs_processed = 0

    def job_update(self, data):
        ''' only requirement from web3py side is to be able to accept a (named) tuple '''
        self.jobs_processed += 1
        if data[1]:
            pass  # do something with successes
        else:
            pass  # do something with failures including TimeoutError

    @property
    def done(self):
        if len(self.jobs_processed) == self.jobs_processed:
            return True
        else:
            return False

    @property
    def progress(self):
        return {'batch-size': self.batch_size, 'processed_jobs': self.jobs_processed}

```

**Comment by pipermerriam** [comment]
@boneyard93501 echoing what @voith said in another issue. 

1. I think `asyncio` is going to be our *goto* for concurrency (and conversely, I'm hesitant to add any thread based concurrency for that reason).
2. I don't have a specific concern here, but I know historically that users who've tried to use web3.py in conjunction with threads have had issues, so a thread based solution seems likely to have unexpected problems.

**Comment by boneyard93501** [comment]
k. converting above to asyncio Queue is fairly trivial. although you'll want to make threaded allowances (concurrent.futures, ThreadPoolExecutor). for event loop registration purposes, i was trying to find the (repo for the) async API  but came up short.  where can i find it? thx.

now, wrapping providers is nice and may have some benefits but the real benefit of batch processing comes from muxing over websocket(s). how do you feel about being opinionated and a) provide a ws-based batch processing solution only (once a suitable ws implementation is in place)  and b) provide an example(s) in the documentation on how to wrap providers right now?

**Comment by jakublipinski** [comment]
@pipermerriam @boneyard93501 I compared the execution time between requesting the transaction receipts for a particular block synchronously and asynchronously. I used `asyncio` and `aiohttp` module for async HTTP requests. I based my code on [the excellent article](https://pawelmhm.github.io/asyncio/python/aiohttp/2016/04/22/asyncio-aiohttp.html) by @pawelmhm. 
The results are very promising. When the program is run locally on a machine with a running parity node I get:
```
sync: 1.843s
async: 0.264s
```
When it's run over the Internet the results are:
```
sync: 8.701s
async: 0.757s
```
Overall I get 7x-11x execution time improvement if the requests are batched and sent asynchronously. My code below:
```python
import timeit
import asyncio

from aiohttp import ClientSession

from web3.providers.base import JSONBaseProvider
from web3.providers import HTTPProvider
from web3 import Web3

# synchronously request receipts for given transactions
def sync_receipts(web3, transactions):
    for tran in transactions:
        web3.eth.getTransactionReceipt(tran)

# asynchronous JSON RPC API request
async def async_make_request(session, url, method, params):
    base_provider = JSONBaseProvider()
    request_data = base_provider.encode_rpc_request(method, params)
    async with session.post(url, data=request_data,
                        headers={'Content-Type': 'application/json'}) as response:
        content = await response.read()
    response = base_provider.decode_rpc_response(content)
    return response

async def run(node_address, transactions):
    tasks = []

    # Fetch all responses within one Client session,
    # keep connection alive for all requests.
    async with ClientSession() as session:
        for tran in transactions:
            task = asyncio.ensure_future(async_make_request(session, node_address,
                                                            'eth_getTransactionReceipt',[tran.hex()]))
            tasks.append(task)

        responses = await asyncio.gather(*tasks)

if __name__ == "__main__":
    eth_node_address = "http://localhost:8545"
    web3 = Web3(HTTPProvider(eth_node_address))

    block = web3.eth.getBlock(web3.eth.blockNumber)
    transactions = block['transactions']

    start_time = timeit.default_timer()
    sync_receipts(web3, transactions)
    print('sync: {:.3f}s'.format(timeit.default_timer() - start_time))

    start_time = timeit.default_timer()
    loop = asyncio.get_event_loop()
    future = asyncio.ensure_future(run(eth_node_address, transactions))
    loop.run_until_complete(future)
    print('async: {:.3f}s'.format(timeit.default_timer() - start_time))
```

**Comment by pipermerriam** [comment]
Yes, these look like very nice performance boosts.  I'd be excited to see someone start working on the `web3.async` API so that users can start taking advantage of the performance gains.

**Comment by medvedev1088** [comment]
@jakublipinski Have you tried batch JSON RPC requests https://www.jsonrpc.org/specification#batch?

**Comment by voith** [comment]
Looks like I had a totally incorrect understanding of this feature. I was thinking of batching in terms of concurrency, but after reading the `json-rpc` documentation, batching seems to mean packing several request objects into a single request.



rpc call Batch:
```
--> [
        {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
        {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},
        {"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},
        {"foo": "boo"},
        {"jsonrpc": "2.0", "method": "foo.get", "params": {"name": "myself"}, "id": "5"},
        {"jsonrpc": "2.0", "method": "get_data", "id": "9"} 
    ]
<-- [
        {"jsonrpc": "2.0", "result": 7, "id": "1"},
        {"jsonrpc": "2.0", "result": 19, "id": "2"},
        {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null},
        {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found"}, "id": "5"},
        {"jsonrpc": "2.0", "result": ["hello", 5], "id": "9"}
    ]
```
Thanks  @medvedev1088

**Comment by Uxio0** [comment]
@medvedev1088 @voith nice! That's what I meant from the beginning. I got a really good performance doing that

**Comment by medvedev1088** [comment]
@Uxio0 @voith Any ideas how it can be added to web3.py? I'm afraid it will require substantial refactoring.

**Comment by Uxio0** [comment]
@medvedev1088 I would suggest to use the same approach as web3 js, https://github.com/ethereum/wiki/wiki/JavaScript-API#batch-requests

Reading the source code I understand it would need a lot of refactor, though

**Comment by voith** [comment]
@Uxio0 @medvedev1088 One thing that is at the top of my mind is that the middlewares will need a some refactoring. Middlewares at the moment work under the assumption that there will always be one method. This feature will need middlewares to accept list of methods. 
> if we build it on top of #657 since everything under web3.async.* will be coroutine based and thus, batching them to run concurrently 

@pipermerriam It doesn't seem like we need concurrency to get batching working(https://github.com/ethereum/web3.py/issues/832#issuecomment-414175168).  All we need is support for packing multiple request objects into a single request.

cc @boneyard93501

**Comment by pipermerriam** [comment]
> It doesn't seem like we need concurrency to get batching work

@voith yes, I was reading batching to mean something different and have since learned about the batching of API requests.

Yes, this is likely to require some underlying architecture changes.  I wouldn't be surprised if we ended up needing a major version bump to really fit this in (though we can add it as an opt-in feature ahead of that schedule or something).

In general, supporting this in web3.py is something I can get behind, it'll just take sorting out how to do it in cleanly.

**Comment by jakublipinski** [comment]
I updated [my test code](https://gist.github.com/jakublipinski/4f1dd698e2bdefdba53d8c69eba3da80) to include [batch requests](https://www.jsonrpc.org/specification#batch) and it turns out batching is indeed faster than synchronous requests but it's still 2.5x slower than asynchronous approach :
```
sync: 80.447s
async: 17.552s
batch: 52.311s
```
(Tested by requesting receipts from the `parity` node from last 20 blocks. The script was run on the same machine)

I looked at the web3.py code and confirm that it would be much easier to introduce batching / async requests if middlewares are refactored so that there is more control on when and how they're called.

**Comment by jakublipinski** [comment]
Test results with Infura mainnet (requesting all receipts from last 5 blocks):
```
sync: 96.470s
async: 3.491s
batch: 3.467s
```
Async requests make sense only if there are several of them. Batching allows you to combine all the requests into one. There will be no benefit from using both batching and async.

**Comment by medvedev1088** [comment]
@jakublipinski I meant the situation when you need to export many blocks. Imagine you want to export 6M blocks. You can't possible batch them into a single request. What you can do thought is to create batches of 100 blocks then run them in parallel. 

This approach is used here https://github.com/medvedev1088/ethereum-etl to export the entire Ethereum blockchain, including blocks, transactions, receipts, logs, ERC20 transfers, contracts, token data.

**Comment by jakublipinski** [comment]
@medvedev1088 Yes, you can use async requests in such case.

**Comment by pipermerriam** [comment]
I'm curious to hear other people's ideas on how to implement this.  Here is the best I've thought of thus far.

First, we need to modify how web3 does requests to support the concept of *building* a request.  I'm thinking that all of the *callable*s that current issue requests get a new api like the following.

```python
>>> w3.eth.getBalance.build_request('0x1234abcd...')
Request(['eth_getBalance', ['0x1234abcd...']])
```

Then we can do something like this.

```python
>>> w3.batch_request(
...     w3.eth.getBalance('0x1234abcd...'),
...     w3.eth.getCode('0x1234abcd...'),
...     ...
... )
...
(
    80000000000  # result of `getBalance`
    '0x63deadbeef'  # result of `getCode`
    ...
)
```

Updates will also have to happen for how middlewares process requests and responses but I think this will be doable by adding a new `Manager.batch_request_blocking` which handles the different logic.

**Comment by Uxio0** [comment]
Hi and sorry for the late response, as this was something needed for us I implemented batching support for contract calls (just on HTTPProvider). Do you think this can be integrated on Web3.py easily?

https://github.com/gnosis/gnosis-py/blob/master/gnosis/eth/ethereum_client.py#L898

So for example for fetching information about an erc20 contract:

```
erc20_contract = w3.eth.contract(erc20_address, abi=erc20_abi)
name, symbol, decimals = ethereum_client.batch_call([
                erc20_contract.functions.name(),
                erc20_contract.functions.symbol(),
                erc20_contract.functions.decimals()
            ])
```

**Comment by pipermerriam** [comment]
Just noting that this fits nicely with the new `Method` class that is used to implement the various APIs.  I think we now have a much firmer foundation to support this functionality if anyone wants to try to take a stab at it.

**Comment by LefterisJP** [comment]
Hey guys any news on this one? Any plans to implement it?

Kind of sucks to have to use the [multicall contract](https://etherscan.io/address/0xeefBa1e63905eF1D7ACbA5a8513c70307C1cE441#code) to do batch queries on the node when the nodes itself allows batch queries.

The js guys already have this functionality for years now and are gonna be making fun of us pythonistas :snake:

**Comment by Pet3ris** [comment]
Couple of thoughts:

- First, I really like the proposal and have an immediate need for this feature
- I think @pipermerriam proposal makes a lot of sense
- I wonder what the implications are for exception handling, how easy will it be to disambiguate which request was problematic
- Finally, I'm relying on `make_request` for request forwarding so would be nice if that didn't break during the refactor :)

**Comment by kclowes** [comment]
@Pet3ris or anyone else on here - if you need this and want to get it rolling, I can definitely help out! I want to finish up a few other things (primarily getting async off the ground) before I put this on my immediate queue, but I can definitely provide feedback and review for anyone wanting to do the legwork now.

> I wonder what the implications are for exception handling, how easy will it be to disambiguate which request was problematic

I think this will be a challenge, but I like @Uxio0's implementation where you pass in the function name, and a param to either raise the exception or not. I think that provides a good start and we can iterate from there. 

> Finally, I'm relying on make_request for request forwarding so would be nice if that didn't break during the refactor :)

Yeah, let's not break `make_request`! I think it would be better to add a new method like `make_batch_request` or something.

**Comment by carver** [comment]
> I think this will be a challenge, but I like @Uxio0's implementation where you pass in the function name, and a param to either raise the exception or not. I think that provides a good start and we can iterate from there.

Yeah, that plus Piper's seems like a reasonable combo to me.

It seems okay to just allow any exceptions to bubble up normally, though maybe we could wrap it in an exception that identifies the index of the call that raised it? Someday there will be something Exception Groups ([PEP 654](https://www.python.org/dev/peps/pep-0654/)), but until then we can do the dumb simple thing.

**Comment by apljungquist** [comment]
If it's of any use I'd be happy to clean up this stop-gap solution that I hacked together. It does not expose any async interfaces but it seems to enable the websocket provider to be used with a threadpool executor thus allowing multiple rpcs to be in flight without the overhead of making individual http requests.

```python
import os
import asyncio
import concurrent.futures
import contextlib
import json
import threading
import time
from typing import Any, Optional, Union

import web3
from eth_typing import URI
from web3 import HTTPProvider
from web3.providers.websocket import (
    DEFAULT_WEBSOCKET_TIMEOUT,
    PersistentWebSocket,
    WebsocketProvider,
)
from web3.types import RPCResponse, RPCEndpoint


class Multiplexer:
    @classmethod
    async def new(cls, conn: PersistentWebSocket):
        return cls(await conn.__aenter__(), asyncio.Queue())

    def __init__(self, conn, queue):
        self._conn = conn
        self._queue = queue
        self._events = {}
        self._responses = {}

    async def send(self):
        while True:
            data = await self._queue.get()
            await self._conn.send(data)

    async def recv(self):
        while True:
            data = await self._conn.recv()
            id = json.loads(data)["id"]  # TODO: error handling?
            self._responses[id] = data
            self._events.pop(id).set()

    async def run(self):
        await asyncio.gather(
            self.send(),
            self.recv(),
        )

    def __call__(self, data):
        event = threading.Event()
        id = json.loads(data)["id"]
        self._events[id] = event
        self._queue.put_nowait(data)
        event.wait()
        return self._responses.pop(id)


class MultiplexingWebsocketProvider(WebsocketProvider):
    def __init__(
        self,
        endpoint_uri: Optional[Union[URI, str]] = None,
        websocket_kwargs: Optional[Any] = None,
        websocket_timeout: int = DEFAULT_WEBSOCKET_TIMEOUT,
    ) -> None:
        super().__init__(endpoint_uri, websocket_kwargs, websocket_timeout)
        self._multiplexer = None
        self._multiplexer_fut = None

    def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
        request_data = self.encode_rpc_request(method, params)
        if self._multiplexer is None:
            assert self._multiplexer_fut is None
            self._multiplexer = asyncio.run_coroutine_threadsafe(
                Multiplexer.new(self.conn),
                self._loop,
            ).result()
            self._multiplexer_fut = asyncio.run_coroutine_threadsafe(
                self._multiplexer.run(),
                self._loop,
            )  # TODO: stop properly
        return json.loads(self._multiplexer(request_data))


@contextlib.contextmanager
def timer():
    t0 = time.perf_counter()
    yield
    t1 = time.perf_counter()
    print(t1 - t0)


def stress(n):
    w3 = web3.Web3(MultiplexingWebsocketProvider(os.environ["ETH_NODE_WSS"]))

    # some dummy load that uses the RPC interfaces
    def call(_):
        nonlocal w3
        return w3.eth.block_number

    with concurrent.futures.ThreadPoolExecutor(max_workers=100) as executor:
        list(executor.map(call, range(n)))


if __name__ == "__main__":
    with timer():
        stress(200)

```

**Comment by rockkoca** [comment]
> If it's of any use I'd be happy to clean up this stop-gap solution that I hacked together. It does not expose any async interfaces but it seems to enable the websocket provider to be used with a threadpool executor thus allowing multiple rpcs to be in flight without the overhead of making individual http requests.
> 
> ```python
> import os
> import asyncio
> import concurrent.futures
> import contextlib
> import json
> import threading
> import time
> from typing import Any, Optional, Union
> 
> import web3
> from eth_typing import URI
> from web3 import HTTPProvider
> from web3.providers.websocket import (
>     DEFAULT_WEBSOCKET_TIMEOUT,
>     PersistentWebSocket,
>     WebsocketProvider,
> )
> from web3.types import RPCResponse, RPCEndpoint
> 
> 
> class Multiplexer:
>     @classmethod
>     async def new(cls, conn: PersistentWebSocket):
>         return cls(await conn.__aenter__(), asyncio.Queue())
> 
>     def __init__(self, conn, queue):
>         self._conn = conn
>         self._queue = queue
>         self._events = {}
>         self._responses = {}
> 
>     async def send(self):
>         while True:
>             data = await self._queue.get()
>             await self._conn.send(data)
> 
>     async def recv(self):
>         while True:
>             data = await self._conn.recv()
>             id = json.loads(data)["id"]  # TODO: error handling?
>             self._responses[id] = data
>             self._events.pop(id).set()
> 
>     async def run(self):
>         await asyncio.gather(
>             self.send(),
>             self.recv(),
>         )
> 
>     def __call__(self, data):
>         event = threading.Event()
>         id = json.loads(data)["id"]
>         self._events[id] = event
>         self._queue.put_nowait(data)
>         event.wait()
>         return self._responses.pop(id)
> 
> 
> class MultiplexingWebsocketProvider(WebsocketProvider):
>     def __init__(
>         self,
>         endpoint_uri: Optional[Union[URI, str]] = None,
>         websocket_kwargs: Optional[Any] = None,
>         websocket_timeout: int = DEFAULT_WEBSOCKET_TIMEOUT,
>     ) -> None:
>         super().__init__(endpoint_uri, websocket_kwargs, websocket_timeout)
>         self._multiplexer = None
>         self._multiplexer_fut = None
> 
>     def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
>         request_data = self.encode_rpc_request(method, params)
>         if self._multiplexer is None:
>             assert self._multiplexer_fut is None
>             self._multiplexer = asyncio.run_coroutine_threadsafe(
>                 Multiplexer.new(self.conn),
>                 self._loop,
>             ).result()
>             self._multiplexer_fut = asyncio.run_coroutine_threadsafe(
>                 self._multiplexer.run(),
>                 self._loop,
>             )  # TODO: stop properly
>         return json.loads(self._multiplexer(request_data))
> 
> 
> @contextlib.contextmanager
> def timer():
>     t0 = time.perf_counter()
>     yield
>     t1 = time.perf_counter()
>     print(t1 - t0)
> 
> 
> def stress(n):
>     w3 = web3.Web3(MultiplexingWebsocketProvider(os.environ["ETH_NODE_WSS"]))
> 
>     # some dummy load that uses the RPC interfaces
>     def call(_):
>         nonlocal w3
>         return w3.eth.block_number
> 
>     with concurrent.futures.ThreadPoolExecutor(max_workers=100) as executor:
>         list(executor.map(call, range(n)))
> 
> 
> if __name__ == "__main__":
>     with timer():
>         stress(200)
> ```

It's great! But, if you increase the n to 400+ for stress function, it might hang forever.

**Comment by Uxio0** [comment]
As @rockkoca says, when increasing the number of requests to the nodes it can hang forever or even hang the node itself.

I've found out that `batch_requests` are not very well handled by the nodes, and there's even some node services that don't enable batch requests. For our use case I decided to implement [multicall](https://github.com/makerdao/multicall) support and it's really really fast compared to doing `calls` using `batch_requests` (orders of magnitude).

For reference, here's the implementation: https://github.com/gnosis/gnosis-py/blob/master/gnosis/eth/multicall.py and [here the guide on how to use it](https://gnosis-py.readthedocs.io/en/latest/quickstart.html#gnosis-eth)

**Comment by sandybradley** [comment]
https://github.com/Narasimha1997/aio-eth

Nice lib implements batch and concurrent requests

**Comment by Fooyao** [comment]
Five years on and still no progress on this issue.

**Comment by shaojiankui** [comment]
need this future

**Comment by Uxio0** [comment]
If you need this feature right now, `safe-eth-py` supports batching and multicall: https://safe-eth-py.readthedocs.io/en/latest/quickstart.html#gnosis-eth

**Comment by dreaded369** [comment]
> If you need this feature right now, `safe-eth-py` supports batching and multicall: https://safe-eth-py.readthedocs.io/en/latest/quickstart.html#gnosis-eth

I can see how 'safe-eth-py' can batch contract calls, but can it batch web3 calls like get_block?

**Comment by Uxio0** [comment]
> I can see how 'safe-eth-py' can batch contract calls, but can it batch web3 calls like get_block?

There's not an API to batch different type of requests, but indeed there's a `get_blocks` method allowing you to get multiple block in the same request.

**Comment by acemasterjb** [comment]
Hello,

would anyone have a solution for performing batch operations where the `block_identifier` is parameterized?

E.g., for 

```python
# Get the results of past calls to Chainlink EACAggregatorProxy.latestAnswer
latest_answer = chainlink_price_feed_contract.functions.latestAnswer().call(
    block_identifier=block_identifier
)
```

I would like to give a different `block_identifier` for a given range of block heights/numbers. Is this possible with `batch_requests`?

I guess I'm a bit confused about what qualifies as `Method[Callable[..., Any]]` which is accepted by `batch_request().add`.
